### ISSUE TITLE
Task Tree Clustering

### ISSUE THREAD

**Issue by yoomlam** [department-of-veterans-affairs/caseflow#12645]
## Background/context
Several problems are being addressed:
1. Identifying failure states (i.e., invalid task trees) for an appeal (e.g., when diagnosing problems for ACD or Bat Team);
2. Documenting valid task trees for understanding how an appeal got to its current state;
3. Automating the above to minimize long-term maintenance costs.

The biggest direct beneficiaries are engineers who must understand and diagnose problems when an appeal gets into a bad state. 

## Goals
The goal is to automatically cluster appeals based on similar task tree structure. Each cluster roughly represents a step in the appeals process, i.e. a cluster contains similar task tree structures. Hence, we can use the information (individual task trees) from a cluster to document that step. We'll query Caseflow's DB for task structures, cluster them, and analyze the clusters. We'll experiment with different clustering parameters such that some clusters contain mostly invalid task trees. Such clusters can be used to alert on appeals that move into a bad state. 
**Extra credit 1**: If we can assign a partial ordering to the clusters, then we can guess how a given task tree likely got to its current state.
**Extra credit 2**: If we associate remedies for each bad-state cluster, then we can offer suggested remedies when a bad state is detected. 

## Success criteria

* Have rich and up-to-date documentation on task trees using realistic examples with no PII using a mostly automated process. I dub this "Data-Driven Documentation" (#12666).
* Likely failure states of appeals are automatically detected. 

## Stakeholders

* Documentation is useful for new hires to learn and engineers to diagnose problems
* Engineers can quickly assess and resolve known failure states by following remedies associated with the states.

## Requirements/stories

1. An appeal's state can be represented by its task tree.
  a. **Assumption**: A task tree with task types, assigned_to_type, and status is sufficient to determine the validity of an appeal. No PII is required to assess the validity of an appeal's state. 
2. Dump all task trees in a canonical representation (i.e., single-line JSON per appeal) into a file(s).
  a. **Assumption**: It is acceptable to transfer files with no PII from the production environment to our local computers for temporary data processing.
3. Cluster the task trees
  a. Create a distance metric for clustering: edit distance (json-diff) or [other tree distance metrics](https://cran.r-project.org/web/packages/Quartet/vignettes/Tree-distance-metrics.pdf).
  b. Experiment with different clustering algorithms.
4. Identify and summarize clusters
  a. Associate each cluster with a step in the appeal process.
  b. Annotate invalid task trees within clusters. We could also write constraint checks agains the task trees much like existing data integrity checkers.
  c. Create associated documentation. Task tree visualization can be automated by translating the JSON task trees into diagramming syntax for input to [PlantUML tool](http://www.plantuml.com/plantuml) for example.
5. Automate the previous steps for long-term maintenance of documentation and clustering models.

The above procedure can be implemented in a few days or can stretch out across several sprints, depending on how precise and accurate we want the results to be.
* I propose we do a quick-and-dirty prototype to assess the utility of the results and decide from there. A shortcut is to represent task trees in its most concise form such that many appeals consolidate into a single task tree. As a result, there will be fewer data points to cluster and fewer to manually assess. We should also incorporate into the distance metric factors that are known to imply an error state. This will result in separating valid and invalid task trees, and/or clustering invalid task trees together. See #12665.

## Open questions

- How many clusters will there be to examine? There may be too many states/clusters to assess.
- Will there be useful cluster? Depends highly on the distance metric.
- How should we deal with old task trees with obsolete/defunct tasks and structure? They were valid in the past but are not relevant or compatible with current task trees.
- How to deal with Intake tasks?
- How to deal with non-AMA appeals?

### Business value
More efficient/effective diagnosis of erroneous appeal state, thereby unblocking and fixing end-users problems faster.

**Comment by yoomlam** [comment]
Closing due to very low priority and impact.
